### PR TITLE
 ta: pkcs11: Add support for more HMAC mechanisms

### PR DESCRIPTION
**Pull request by vchong**

### CODE REVIEW

**Comment by vchong** [comment]
Relates to:
- https://github.com/OP-TEE/optee_os/issues/4283

File: ta/pkcs11/src/processing_symm.c [hunk 1]
```diff
@@ -612,6 +676,9 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 	uint32_t in2_size = 0;
 	bool output_data = false;
 	struct active_processing *proc = session->processing;
+	static uint32_t hmac_len;
+	uint8_t computed_mac[TEE_MAX_HASH_SIZE];
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:675]: `= { 0 }`

File: ta/pkcs11/src/processing_symm.c [hunk 2]
```diff
@@ -751,6 +824,70 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+	case PKCS11_CKM_MD5_HMAC_GENERAL:
+	case PKCS11_CKM_SHA_1_HMAC_GENERAL:
+	case PKCS11_CKM_SHA224_HMAC_GENERAL:
+	case PKCS11_CKM_SHA256_HMAC_GENERAL:
+	case PKCS11_CKM_SHA384_HMAC_GENERAL:
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		if (session->processing->extra_ctx) {
+			hmac_len = *(uint32_t *)session->processing->extra_ctx;
+			/*
+			 * remove ptr to NW addr so that it doesn't get
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:831]: this is ugly (sorry). better allocate a 32bit in `extra_ctx` and use it.

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "ta: pkcs11: Add support for more HMAC mechanisms".

File: ta/pkcs11/src/processing_symm.c [hunk 3]
```diff
@@ -461,6 +497,34 @@ init_tee_operation(struct pkcs11_session *session,
...
+		}
+
+		/*
+		 * store output length in bytes (CK_MAC_GENERAL_PARAMS) in
+		 * extra_ctx
+		 */
+		session->processing->extra_ctx = (void *)proc_params->data;
+
+		DMSG("hmac_len = %u\n",
+		     *(uint32_t *)session->processing->extra_ctx);
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c:521]: We should add some checks on hmac_len. This passed length should be in the range from 0 and output length based on SHA in the mechanism.

eg for CKM_SHA512_224_HMAC_GENERAL, length of the output should be in the range 0-28, so the hmac_len can't be more than 28.

File: ta/pkcs11/src/processing_symm.c [hunk 4]
```diff
@@ -751,6 +824,70 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+			DMSG("hmac_len = %u", hmac_len);
+		}
+
+		switch (function) {
+		case PKCS11_FUNCTION_SIGN:
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size, out_buf,
+						  &out_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
```
Comment (ruchi393) [inline ta/pkcs11/src/processing_symm.c:845]: Probably some error check if hmac_len > out_sz

Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c:845]: Another case could be :

User passes output buffer of size -> hmac_len (which is of size less than the expected out_sz) . However, TEE_MACComputeFinal will return short buffer error in this case in current flow. We will need to take care of this.

eg case:
For hmac_sha256_general case, caller passes hmac_len as 10 bytes and output_buffer also of out_sz 10 bytes. In the current flow, this will return error to the caller  which is not right. You might want to use your own temporary buffer here.

File: ta/pkcs11/src/processing_symm.c [hunk 5]
```diff
@@ -461,6 +497,34 @@ init_tee_operation(struct pkcs11_session *session,
...
+		if (!proc_params->data) {
+			EMSG("Unexpected null pointer");
+			return PKCS11_CKR_MECHANISM_PARAM_INVALID;
+		}
+
+		/*
+		 * store output length in bytes (CK_MAC_GENERAL_PARAMS) in
+		 * extra_ctx
+		 */
+		session->processing->extra_ctx = (void *)proc_params->data;
```
[the remark this answers is not in the document]
Comment (vesajaaskelainen) [reply ta/pkcs11/src/processing_symm.c:518]: Or even a struct that has more meaningful field name.

File: ta/pkcs11/src/processing_symm.c [hunk 6]
```diff
@@ -751,6 +819,70 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+			output_data = true;
+			rc = tee2pkcs_error(res);
+			break;
+		case PKCS11_FUNCTION_VERIFY:
+			/* must compute full mac before comparing partial */
+			res = TEE_MACComputeFinal(proc->tee_op_handle, in_buf,
+						  in_size, computed_mac,
+						  &computed_mac_size);
+
+			DMSG("computed_mac_size = %u\n", computed_mac_size);
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:857]: This is development time debug and should be removed.

File: ta/pkcs11/src/processing_symm.c [hunk 7]
```diff
@@ -612,6 +671,9 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 	uint32_t in2_size = 0;
 	bool output_data = false;
 	struct active_processing *proc = session->processing;
+	static uint32_t hmac_len;
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:674]: why static here?

**Review by vchong — COMMENTED** [review]
Updated

**Review by etienne-lms — COMMENTED** [review]
Commit "ta: pkcs11: Fix typo":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` 

Commit: "ta: pkcs11: Fix sign size comparison":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with commit log header line updated to something like: "ta: pkcs11: Fix HMAC input signature size comparison".

Comments for commit "ta: pkcs11: Add support for more HMAC mechanisms".

File: ta/pkcs11/src/processing_symm.c [hunk 8]
```diff
@@ -751,6 +858,62 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 
 		break;
 
+	case PKCS11_CKM_MD5_HMAC_GENERAL:
+	case PKCS11_CKM_SHA_1_HMAC_GENERAL:
+	case PKCS11_CKM_SHA224_HMAC_GENERAL:
+	case PKCS11_CKM_SHA256_HMAC_GENERAL:
+	case PKCS11_CKM_SHA384_HMAC_GENERAL:
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		if (session->processing->extra_ctx)
```
[the remark this answers is not in the document]
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c]: ok

File: ta/pkcs11/src/processing_symm.c [hunk 9]
```diff
@@ -751,6 +858,62 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+		switch (function) {
+		case PKCS11_FUNCTION_SIGN:
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size, out_buf,
+						  &out_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
+				out_size = hmac_len;
+			}
+			output_data = true;
```
[the remark this answers is not in the document]
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c]: You're right, thanks.

File: ta/pkcs11/src/processing_symm.c [hunk 10]
```diff
@@ -751,6 +858,62 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+
+			if (res == TEE_SUCCESS) {
+				/*
+				 * Only the first in2_size bytes of the
+				 * signature to be verified is passed in from
+				 * NW
+				 */
+				if (TEE_MemCompare(in2_buf, computed_mac,
+						   in2_size)) {
+					EMSG("C_VerifyFinal() MAC mismatch");
```
[the remark this answers is not in the document]
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c]: ok

**Comment by vchong** [comment]
Updated

File: ta/pkcs11/src/processing_symm.c (hunk 4 above)
Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c:876]: For PKCS11_CKM_SHA512_HMAC_GENERAL usecases, caller can request an HMAC length of the output  in the range 0-28. In this case , if the length expected by caller is say 14 bytes, I think it should not be a short buffer error for the caller to pass an output buffer of size 14 bytes to the implementation.

Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:876]: > For PKCS11_CKM_SHA512_HMAC_GENERAL usecases, caller can request an HMAC length of the output in the range 0-28. In this case , if the length expected by caller is say 14 bytes, I think it should not be a short buffer error for the caller to pass an output buffer of size 14 bytes to the implementation.

Ok, I think I get what you mean. Let me try to do some refactoring.

Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:884]: @ruchi393 The refactoring turned out to be less work than previously imagined. Please take a look and comment. Thanks!

File: lib/libutee/tee_api_operations.c [hunk 11]
```diff
@@ -1242,7 +1242,7 @@ TEE_Result TEE_MACCompareFinal(TEE_OperationHandle operation,
 			       const void *mac, uint32_t macLen)
 {
 	TEE_Result res;
-	uint8_t computed_mac[TEE_MAX_HASH_SIZE];
+	uint8_t computed_mac[TEE_MAX_HASH_SIZE] = { 0 };
```
Comment (ruchi393) [inline lib/libutee/tee_api_operations.c:1245]: This is change in generic libutee. Should be part of separate commit.

**Comment by ruchi393** [comment]
@vchong , can you please squash the changes and re-submit the PR.

**Comment by vchong** [comment]
@ruchi393 @etienne-lms @vesajaaskelainen Rebased and squashed. Please approve if ok. Thanks.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "lib: libutee: init array to 0"
* "ta: pkcs11: Fix typo"
* "ta: pkcs11: Fix sign size comparison"

Regarding commit "ta: pkcs11: Change error return code":
do we need this change? Are we expected to reach `input_sign_size_is_valid()` with an invalid mechanism ID. If not, I think the general error (we could aslo panic) is preferred.

Some comments for commit "ta: pkcs11: Add support for more HMAC mechanisms".

**Comment by vchong** [comment]
> Regarding commit "ta: pkcs11: Change error return code":
> do we need this change? Are we expected to reach input_sign_size_is_valid() with an invalid mechanism ID. If not, I think the general error (we could aslo panic) is preferred.

Ok. I'll remove the commit.

> Some comments for commit "ta: pkcs11: Add support for more HMAC mechanisms".

@etienne-lms I don't see the comments. You might have missed posting them?

**Review by etienne-lms — COMMENTED** [review]
> @etienne-lms I don't see the comments. You might have missed posting them?

Right, here they are. Sorry.

File: ta/pkcs11/src/processing_symm.c [hunk 12]
```diff
@@ -751,6 +858,79 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+				/* truncate to hmac_len */
+				memcpy(out_buf, computed_mac, hmac_len);
+			}
+			if (res == TEE_SUCCESS ||
+			    res == TEE_ERROR_SHORT_BUFFER)
+			    output_data = true;
+
+			/* inform client of required size */
+			out_size = hmac_len;
+			rc = tee2pkcs_error(res);
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply ta/pkcs11/src/processing_symm.c:899]: Discard my comment. As you pointed me libckteec does that job.
Using a generic `rc = tee2pkcs_error(res);` is fine here.

File: ta/pkcs11/src/processing_symm.c (hunk 4 above)
Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c:890]: LGTM

**Review by vchong — COMMENTED** [review]
All comments addressed

File: ta/pkcs11/src/processing_symm.c [hunk 13]
```diff
@@ -751,6 +857,78 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		assert(session->processing->extra_ctx);
+		hmac_len = *(uint32_t *)session->processing->extra_ctx;
+
+		switch (function) {
+		case PKCS11_FUNCTION_SIGN:
+			if (out_size && out_size < hmac_len) {
+				/* inform client of required size */
+				out_size = hmac_len;
+				return PKCS11_CKR_BUFFER_TOO_SMALL;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:874]: Issue the output size not returned.
```patch
-				return PKCS11_CKR_BUFFER_TOO_SMALL;
+				output_data = true;
+				rc = PKCS11_CKR_BUFFER_TOO_SMALL;
+				goto out;

Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:874]: Ok. With this change, I was able to simplify the code a bit as well. Please see if the following changes are acceptable. Thanks.

Comment (etienne-lms) [reply ta/pkcs11/src/processing_symm.c]: LGTM

**Comment by vchong** [comment]
Rebased and tag added. Thanks!

@ruchi393 @vesajaaskelainen Would you like to add your tag as well?

File: ta/pkcs11/src/processing_symm.c [hunk 14]
```diff
@@ -461,6 +533,31 @@ init_tee_operation(struct pkcs11_session *session,
...
+		pkcs11_data = TEE_Malloc(sizeof(uint32_t),
+					 TEE_MALLOC_FILL_ZERO);
+		if (!pkcs11_data)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		memcpy(pkcs11_data, proc_params->data, sizeof(uint32_t));
+
+		rc = input_hmac_len_is_valid(proc_params, *pkcs11_data);
+		if (rc)
+			return rc;
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:557]: we leak memory here (pkcs11_data) if the len is not valid.

File: lib/libutee/tee_api_operations.c (hunk 11 above)
Comment (vesajaaskelainen) [reply lib/libutee/tee_api_operations.c:1245]: Wondering if this is a good change or not. In theory there should not be a problem as only valid bytes are being accessed due to size checks.

Guideline is to initialize variables to good initial state which says that we should in example zeroize it.

But looking around of this there also other variables that do not get initialized.

@jenswi-linaro & @jforissier should we be concerned? And do batch changes for those in another PR?

As far as I can see this does not have affect for this particular PRs functionality.

File: ta/pkcs11/src/processing_symm.c [hunk 15]
```diff
@@ -751,6 +857,73 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+				EMSG("Invalid signature size: %"PRIu32,
+				     in2_size);
+				return PKCS11_CKR_SIGNATURE_LEN_RANGE;
+			}
+
+			if (res == TEE_SUCCESS) {
+				/*
+				 * Only the first in2_size bytes of the
+				 * signature to be verified is passed in from
+				 * NW
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:910]: NW -- perhaps REE (or be more verbose) as that is commonly used with TEE terminology.

Or just passed from "caller" then if someone does TA->TA call then it would also be correct.

File: ta/pkcs11/src/processing_symm.c [hunk 16]
```diff
@@ -751,6 +857,73 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size,
+						  computed_mac,
+						  &computed_mac_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
+				memcpy(out_buf, computed_mac, hmac_len);
+			}
+			if (res == TEE_SUCCESS || res == TEE_ERROR_SHORT_BUFFER)
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:887]: I suppose TEE_ERROR_SHORT_BUFFER should be quite bad error as we have fixed buffer that should be good enough?

Perhaps that should even be assert/panic?

File: ta/pkcs11/src/processing_symm.c [hunk 17]
```diff
@@ -751,6 +857,73 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+				output_data = true;
+
+			/* inform client of required size */
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:890]: As for logic goes -- there was already step before with exactly same comment to verify and report back required size. Eg. check should be done for size.

If we get here we have more or less copied the data there already.

You could get rid of the comment altogether if you move setting of out_size next to memory copying spot.

Now this comment should reflect that we just tell what we copied?

File: ta/pkcs11/src/processing_symm.c [hunk 18]
```diff
@@ -751,6 +857,73 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+				goto out;
+			}
+
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size,
+						  computed_mac,
+						  &computed_mac_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
+				memcpy(out_buf, computed_mac, hmac_len);
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:885]: TEE_MemMove() is the one we probably should be using here.

File: ta/pkcs11/src/processing_symm.c [hunk 19]
```diff
@@ -461,6 +533,31 @@ init_tee_operation(struct pkcs11_session *session,
...
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		if (proc_params->size != sizeof(uint32_t))
+			return PKCS11_CKR_MECHANISM_PARAM_INVALID;
+
+		pkcs11_data = TEE_Malloc(sizeof(uint32_t),
+					 TEE_MALLOC_FILL_ZERO);
+		if (!pkcs11_data)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		memcpy(pkcs11_data, proc_params->data, sizeof(uint32_t));
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/processing_symm.c:553]: TEE_MemMove() here too

**Comment by vesajaaskelainen** [comment]
For commits:
- https://github.com/OP-TEE/optee_os/pull/4746/commits/5dceda0e1bb7b8df162f47bd2cc66f5830a6546a ta: pkcs11: Fix typo
- https://github.com/OP-TEE/optee_os/pull/4746/commits/0c702785b58db1bcbab097cc390d0362419bcbf8 ta: pkcs11: Fix sign size comparison

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

File: ta/pkcs11/src/processing_symm.c (hunk 16 above)
Comment (etienne-lms) [reply ta/pkcs11/src/processing_symm.c:887]: agree

File: ta/pkcs11/src/processing_symm.c (hunk 17 above)
Comment (etienne-lms) [reply ta/pkcs11/src/processing_symm.c:890]: input value for `out_size` may be higher than `hmac_len` and here we are expected to return the effective output data size. The implementation looks good to me.

File: ta/pkcs11/src/processing_symm.c (hunk 15 above)
Comment (etienne-lms) [reply ta/pkcs11/src/processing_symm.c:910]: Agree with "caller" or even "client".

**Review by etienne-lms — COMMENTED** [review]
Thanks @vesajaaskelainen for all the findings. Few comments.

File: ta/pkcs11/src/processing_symm.c (hunk 14 above)
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:557]: Good catch! Thanks!

File: ta/pkcs11/src/processing_symm.c (hunk 15 above)
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:910]: Changed to "caller"

File: ta/pkcs11/src/processing_symm.c (hunk 16 above)
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:887]: You're right, thanks! I removed `TEE_ERROR_SHORT_BUFFER` since that shouldn't happen and moved setting out `output_data` into the statement above.

File: ta/pkcs11/src/processing_symm.c (hunk 18 above)
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:885]: ok

File: ta/pkcs11/src/processing_symm.c (hunk 19 above)
Comment (vchong) [reply ta/pkcs11/src/processing_symm.c:553]: ok

**Comment by vchong** [comment]
Thanks @vesajaaskelainen and @etienne-lms.

@vesajaaskelainen All comments addressed. Please approve if ok.

File: ta/pkcs11/src/processing_symm.c [hunk 20]
```diff
@@ -550,11 +550,13 @@ init_tee_operation(struct pkcs11_session *session,
 		if (!pkcs11_data)
 			return PKCS11_CKR_DEVICE_MEMORY;
 
-		memcpy(pkcs11_data, proc_params->data, sizeof(uint32_t));
+		TEE_MemMove(pkcs11_data, proc_params->data, sizeof(uint32_t));
 
 		rc = input_hmac_len_is_valid(proc_params, *pkcs11_data);
-		if (rc)
+		if (rc) {
+			TEE_Free((void *)pkcs11_data);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:557]: is the cast really needed?

Comment (vchong) [reply ta/pkcs11/src/processing_symm.c]: Guess not. Removed.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
`Reviewed-by: Ruchika Gupta <ruchika.gupta@linaro.org>`

**Comment by vesajaaskelainen** [comment]
Once fixup commits has been fixed up:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

I assume checkpatch would then pass... Don't know about IBART looks like device issue.

**Comment by vchong** [comment]
Rebased and tags added. Thanks for all the reviews, comments and feedbacks!

@jforissier Please merge along with https://github.com/OP-TEE/optee_test/pull/535! Thanks!
